### DISCUSSION TITLE
Anything wrong with maven-bundle-plugin?

### DISCUSSION THREAD

**Discussion by laeubi** [eclipse-m2e/m2e-core#422] [Q&A] [answered]
The maven bundle-plugin has a [lifecycle-mapping-metadata.xml](
https://github.com/apache/felix-dev/blob/master/tools/maven-bundle-plugin/src/main/resources/META-INF/m2e/lifecycle-mapping-metadata.xml) and from the declaration it should be called on each change (`runOnIncremental=true`).

What I do is the following:

1) go to the target/classes/META-INF and open the MANIFEST.MF in an editor
2) create a new class in a new package (**that should trigger a modified manifest as a new package is exported**)
3) But the MANIFEST.MF file stays unchanged (timestamp remains the same and content as well)
4) now I do `Maven>Update Project` and choose to update the given project, or use `Project>Clean`.
5) now I see my changes applied

Is my expectation wrong or is the maven-bundle-plugin doing something wrong? Is it possible to get some info about invoked mojos?

Use configuration is:

```
<build>
    <plugins>
        <plugin>
            <groupId>org.apache.felix</groupId>
            <artifactId>maven-bundle-plugin</artifactId>
            <version>5.1.2</version>
            <extensions>true</extensions>
            <configuration>
                <instructions>
                    <Export-Package>*</Export-Package>
                </instructions>
            </configuration>
        </plugin>
    </plugins>
</build>
```

**Comment by HannesWell** [comment]
I cannot add much value to this topic, just that I encountered similar behaviour when reworking m2e's maven-runtime components.
But I have admit that I didn't investigate it further.

**Comment by mickaelistria** [comment]
EDIT: s/goal/phase/
I sincerely have no clue, and never investigated this topic. I think you'll have to debug it ;) Some things that may affect the decision of whether to run the goal on incremental build vs full build (which I think "Update Project" does trigger) can be the phases that are mapped to the different kind of builds. Maybe the full build would lookup mojos from all phases while the incremental build would only query a subset of the phases?

**Reply by laeubi** [reply]
> Maybe the full build would lookup mojos from all phases while the incremental build would only query a subset of the phases?

Yeah that's something I'd like to check first if there are any 'obvious' problems, I have never used these m2e connector/extensions so maybe my understanding is wrong here.

I'll try to debug this case today, any hint where in m2e this things happen, I assume starting at the MavenBuilder would be a good choice but if anyone knows from the top of his head a more specific place any hints are welcome :-)

**Reply by mickaelistria** [reply]
I think `MojoExecutionBuildParticipant.build()` is a good starting point to place a breakpoint and start monitoring/diffing the case of full build vs incremental one. The m2e-core-tests also has some tests about that (look for `runOnIncremental` in the code, mostly in `LifeCycleMappingTests`) where you could automate your case, saving lots of clicks.

**Reply by laeubi** [reply]
Right, it seems there is already a [MavenBundlePluginConfigurator](https://github.com/tesla/m2eclipse-tycho/blob/master/org.sonatype.tycho.m2e/src/org/sonatype/tycho/m2e/felix/internal/MavenBundlePluginConfigurator.java) in my workspace (do I have to install this or is it included by default?) what has some magic to not run to often that causing problems here.

Does anyone know more about that  tesla / m2eclipse-tycho @fbricon @ifedorenko ?

Is it no longer maintained? Would it make sense to in-cooperate this either in tycho or m2e (maybe as a separate sub-project?)

**Reply by laeubi** [reply]
Strange enought according to https://github.com/tesla/m2eclipse-tycho/commit/22db07983aa7a92482174c90144eef25e83943d0 for versions >= 3.2.0 this plugin should no longer be executed but I see:
```

AbstractCustomizableLifecycleMapping - Build participants for MavenProject: local.maven:project:0.0.1-SNAPSHOT @ /home/christoph/workspaces/runtime-New_configuration/felix-bundle-project/pom.xml
0AbstractCustomizableLifecycleMapping - Mojo execution key: org.apache.maven.plugins:maven-resources-plugin:3.2.0:resources (execution: default-resources, phase: process-resources)
0AbstractCustomizableLifecycleMapping - 	Action: execute
0AbstractCustomizableLifecycleMapping - Mojo execution key: org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (execution: default-compile, phase: compile)
0AbstractCustomizableLifecycleMapping - Mojo execution key: org.apache.maven.plugins:maven-resources-plugin:3.2.0:testResources (execution: default-testResources, phase: process-test-resources)
0AbstractCustomizableLifecycleMapping - 	Action: execute
0AbstractCustomizableLifecycleMapping - Mojo execution key: org.apache.maven.plugins:maven-compiler-plugin:3.8.1:testCompile (execution: default-testCompile, phase: test-compile)
0AbstractCustomizableLifecycleMapping - Mojo execution key: org.apache.maven.plugins:maven-surefire-plugin:3.0.0-M5:test (execution: default-test, phase: test)
0AbstractCustomizableLifecycleMapping - Mojo execution key: org.apache.felix:maven-bundle-plugin:5.1.2:bundle (execution: default-bundle, phase: package)
0AbstractCustomizableLifecycleMapping - 	Action: configurator
0AbstractCustomizableLifecycleMapping - 		Project configurator : id=maven-bundle-plugin class=org.sonatype.tycho.m2e.felix.internal.MavenBundlePluginConfigurator
0AbstractCustomizableLifecycleMapping - 		Build participant: org.sonatype.tycho.m2e.felix.internal.MavenBundlePluginConfigurator$1
0AbstractCustomizableLifecycleMapping - Mojo execution key: org.apache.maven.plugins:maven-install-plugin:3.0.0-M1:install (execution: default-install, phase: install)
0AbstractCustomizableLifecycleMapping - Mojo execution key: org.apache.felix:maven-bundle-plugin:5.1.2:install (execution: default-install, phase: install)
0AbstractCustomizableLifecycleMapping - Mojo execution key: org.apache.maven.plugins:maven-deploy-plugin:3.0.0-M1:deploy (execution: default-deploy, phase: deploy)
0AbstractCustomizableLifecycleMapping - Mojo execution key: org.apache.felix:maven-bundle-plugin:5.1.2:deploy (execution: default-deploy, phase: deploy)
```

in the logfile, is it still listed if it is not executed?

**Reply by laeubi** [reply]
I also see this:
> Executing build participant org.sonatype.tycho.m2e.felix.internal.MavenBundlePluginConfigurator$1 for plugin execution org.apache.felix:maven-bundle-plugin:5.1.2:bundle (execution: default-bundle, phase: package)

so maybe the root cause is that the filtering of `MavenBundlePluginConfigurator` does not work and thus overrides the default action from inside the bundle plugin?

**Comment by laeubi** [comment]
There are two issues:
- https://github.com/eclipse-m2e/m2e-core/issues/426
- and I have not setup an execution for the "manifest" goal what currently is the only m2e "mapped" goal in felix plugin

For the first I have no idea and for the second I wonder if we can make this more convient to the user? Actually I don't really like to configure special executions in the pom (and don't know if the might interfere with the default build), so any suggestions?

**Comment by laeubi** [comment]
What I have done so far:

1. uninstalled the tycho-m2e connector
2. added an execution binding:
```
<plugin>
	<groupId>org.apache.felix</groupId>
	<artifactId>maven-bundle-plugin</artifactId>
	<version>5.1.2</version>
	<extensions>true</extensions>
	<configuration>
		<instructions>
			<Export-Package>*</Export-Package>
			<Import-Package>!sun.*,*</Import-Package>
		</instructions>
	</configuration>
	<executions>
		<execution>
			<goals>
				<goal>manifest</goal>
			</goals>
			<phase>prepare-package</phase>
		</execution>
	</executions>
</plugin>
```

This at least make the felix plugin to be executed. Question is: can we do anything to not define an execution here? Or is it even desired?

**Comment by laeubi** [comment, marked as answer]
I finally go the solution, the felix bundle plugin requires to explicitly enable the incremental build feature:

```
<plugins>
	<plugin>
		<groupId>org.apache.felix</groupId>
		<artifactId>maven-bundle-plugin</artifactId>
		<version>5.1.2</version>
		<extensions>true</extensions>
		<configuration>
			<instructions>
				<Export-Package>*</Export-Package>
				<Import-Package>!sun.*,*</Import-Package>
			</instructions>
			<supportIncrementalBuild>true</supportIncrementalBuild>
		</configuration>
		<executions>
			<execution>
				<goals>
					<goal>manifest</goal>
				</goals>
				<phase>prepare-package</phase>
			</execution>
		</executions>
	</plugin>
</plugins>
```

I'll mark this as the answer but it is really a bit undesirable from a users POV:

1. I have to get rid of the old tycho-m2e binding (this will hopefully be [fixed soon](https://github.com/takari/m2e-discovery-catalog/pull/70))
2. Set `BUNDLE_ROOT_PATH=target/classes` in `org.eclipse.pde.core.prefs`
3. I have to add an execution binding for the manifest goal
4. I have to explicitly set `supportIncrementalBuild`

These are all steps I suppose the "usual maven developer" won't do and thus might get the feeling it is not reliable working in the IDE. If we really want to have more user acceptance I think this should be changed, not sure if this is already possible with the current connectors aproach:

1. On project update `BUNDLE_ROOT_PATH=target/classes` in `org.eclipse.pde.core.prefs`should be updated to reflect the current output folder
2. `org.eclipse.pde.PluginNature` is added
3. an execution for the `manifest` goal is implicitly added and `supportIncrementalBuild=true` is set if not explicitly disabled.

**Reply by kwin** [reply]
I would recommend switching to bnd-maven-plugin which IMHO has superior m2e integration.

**Reply by laeubi** [reply]
Can you explain what makes it superior (from a users POV)? I also tested with BND (see https://github.com/eclipse-m2e/m2e-core/discussions/444 ) and found it having nothing special beside that I don't need to enable incremental builds.
